_posts/2016-05-05-cargo-pillars.md

```diff
@@ -0,0 +1,489 @@
+project successfully compiles, subsequent compiles will use exactly the same
+source code, across machines and environments**.
+
+To achieve this guarantee, Cargo uses several strategies:
+
+1. The first time a build succeeds, Cargo emits a `Cargo.lock` file, which
+   contains a manifest of precisely which source code was used in the
+   build. (more on "precise" in a bit)
+
+2. Cargo manages the entire workflow, from running tests and benchmarks, to
+   building release artifacts, to running executables for debugging. This allows
+   Cargo to ensure that all dependencies (direct and indirect) are downloaded
+   and properly configured for these use-cases without the user having to do
+   anything extra.
+
+3. Cargo standardizes important environment configuration, like optimization
+   level, static and dynamic linking, and architecture. Combined with the
+   `Cargo.lock`, this makes the results of building, testing and executing Cargo
+   projects highly predictable.
+
+### Predictability By Example
+
+To illustrate these strategies, let's build an example crate using Cargo. To
+keep things simple, we'll create a small `datetime` crate that exposes date and
+time functionality.
+
+First, we'll use `cargo new` to start us out:
+
+```
+$ cargo new datetime
+$ ls
+Cargo.toml src
+$ cat Cargo.toml
+[package]
+name = "datetime"
+version = "0.1.0"
+authors = ["Yehuda Katz <wycats@gmail.com>"]
+
+[dependencies]
+```
+
+We don't want to build the date or time functionality from scratch, so let's
+edit the `Cargo.toml` and add the `time` crate from crates.io:
+
+```
+  [package]
+  name = "datetime"
+  version = "0.1.0"
+  authors = ["Yehuda Katz <wycats@gmail.com>"]
+
+  [dependencies]
++ time = "0.1.35"
+```
+
+Now that we've added the `time` crate, let's see what happens if we ask Cargo to
+build our package:
+
+```
+$ cargo build
+   Compiling winapi v0.2.6
+   Compiling libc v0.2.10
+   Compiling winapi-build v0.1.1
+   Compiling kernel32-sys v0.2.2
+   Compiling time v0.1.35
+   Compiling datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+Whoa! That's a lot of crates. **The biggest part of Cargo's job is to provide
+enough predictability to allow functionality like the `time` crate to be broken
+up into smaller crates that do one thing and do it well**.
+
+Now that we successfully built our crate, what happens if we try to build it again?
+
+```
+$ cargo build
+
+```
+
+Nothing happened at all. Why's that? We can always ask Cargo to give us more
+information through the `--verbose` flag, so let's do that:
+
+```
+$ cargo build --verbose
+       Fresh libc v0.2.10
+       Fresh winapi v0.2.6
+       Fresh winapi-build v0.1.1
+       Fresh kernel32-sys v0.2.2
+       Fresh time v0.1.35
+       Fresh datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+Cargo isn't bothering to recompile packages that it knows are "fresh". If you've
+ever had to configure a `Makefile` before, this should be a breath of fresh air
+(no pun intended).
```

Comment:
Wasn't quite sure what was meant here.  Don't makefiles also support not recompiling when nothing is changed?  At the very least, we may want to tweak wording to be specific about the advantage.